src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs

```diff
@@ -0,0 +1,456 @@
+                    {
+                        SetValue(writer, "version", versionString);
+                    }
+
+                    writer.WriteObjectEnd();
+                }
+                else
+                {
+                    writer.WriteNameValue(dependency.Name, versionString);
+                }
+            }
+
+            writer.WriteObjectEnd();
+        }
+
+        private static void SetImports(IObjectWriter writer, IList<NuGetFramework> frameworks)
+        {
+            if (frameworks?.Any() == true)
+            {
+                var sortedImports = frameworks.Select(framework => framework.GetShortFolderName())
+                                              .OrderBy(name => name, StringComparer.Ordinal);
+
+                writer.WriteNameArray("imports", sortedImports);
+            }
+        }
+
+        private static void SetFrameworks(IObjectWriter writer, IList<TargetFrameworkInformation> frameworks)
+        {
+            if (frameworks.Any())
+            {
+                writer.WriteObjectStart("frameworks");
+
+                var sortedFrameworks = frameworks.OrderBy(framework => framework.FrameworkName.DotNetFrameworkName, StringComparer.Ordinal);
+
+                foreach (var framework in sortedFrameworks)
+                {
+                    writer.WriteObjectStart(framework.FrameworkName.GetShortFolderName());
+
+                    SetDependencies(writer, framework.Dependencies);
+                    SetImports(writer, framework.Imports);
+
+                    writer.WriteObjectEnd();
+                }
+
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private static void SetTools(IObjectWriter writer, IList<ToolDependency> tools)
+        {
+            if (tools.Any())
+            {
+                writer.WriteObjectStart("tools");
+
+                var sortedTools = tools.OrderBy(tool => tool.LibraryRange.Name, StringComparer.Ordinal);
+
+                foreach (var tool in sortedTools)
+                {
+                    writer.WriteObjectStart(tool.LibraryRange.Name);
+                    writer.WriteNameValue("version", tool.LibraryRange.VersionRange.ToNormalizedString());
+
+                    if (tool.Imports.Any())
+                    {
+                        SetImports(writer, tool.Imports);
+                    }
+
+                    writer.WriteObjectEnd();
+                }
+
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private static void SetValue(IObjectWriter writer, string name, string value)
+        {
+            if (value != null)
+            {
+                writer.WriteNameValue(name, value);
+            }
+        }
+
+        private static void SetArrayValue(IObjectWriter writer, string name, IEnumerable<string> values)
+        {
+            if (values != null && values.Any())
+            {
+                values = values.OrderBy(value => value, StringComparer.Ordinal);
```

Comment:
This does not seem like a safe thing to do all the time. Arrays are not necessary collections where order does not matter. The caller of this method should do the sorting on their own because they have the knowledge. Alternatively, a "`SetSortedArrayValue` method could be added.